pkg/provider/ngrok/client.go

```diff
@@ -0,0 +1,237 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or impliec.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ngrok
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/ngrok/ngrok-api-go/v7"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/utils/ptr"
+
+	esv1 "github.com/external-secrets/external-secrets/apis/externalsecrets/v1"
+)
+
+const (
+	defaultDescription = "Managed by External Secrets Operator"
+)
+
+var (
+	errWriteOnlyOperations     = errors.New("not implemented - the ngrok provider supports write-only operations")
+	errVaultDoesNotExist       = errors.New("vault does not exist")
+	errVaultSecretDoesNotExist = errors.New("vault secret does not exist")
+	errCannotPushNilSecret     = errors.New("cannot push nil secret")
+)
+
+type VaultClient interface {
+	Create(context.Context, *ngrok.VaultCreate) (*ngrok.Vault, error)
+	List(*ngrok.Paging) ngrok.Iter[*ngrok.Vault]
+}
+
+type SecretsClient interface {
+	Create(context.Context, *ngrok.SecretCreate) (*ngrok.Secret, error)
+	Update(context.Context, *ngrok.SecretUpdate) (*ngrok.Secret, error)
+	Delete(context.Context, string) error
+	List(*ngrok.Paging) ngrok.Iter[*ngrok.Secret]
+}
+
+type client struct {
+	vaultClient   VaultClient
+	secretsClient SecretsClient
+	vaultName     string
+}
+
+func (c *client) PushSecret(ctx context.Context, secret *corev1.Secret, data esv1.PushSecretData) error {
+	if secret == nil {
+		return errCannotPushNilSecret
+	}
+
+	// First, get the vault by name. If it doesn't exist, create it.
+	vault, err := c.getVaultByName(ctx, c.vaultName)
+	if err != nil {
+		if !errors.Is(err, errVaultDoesNotExist) {
+			return err
+		}
```

Comment:
nit: move out of err block (errors.Is has an inbuilt nil check) to reduce nesting

Reply:
I couldn't find a better way to unnest this due to the negation of `errors.Is`. Open to suggestions here though in case I missed it :man_facepalming:

Reply:
no problem, how about something like this:

```go
vault, err := c.getVaultByName(ctx, c.vaultName)
if errors.Is(err, errVaultDoesNotExist) {
	vault, err = c.vaultClient.Create(ctx, &ngrok.VaultCreate{
		Name:        c.vaultName,
		Description: defaultDescription,
	})
	if err != nil {
		return fmt.Errorf("failed to create vault: %w", err)
	}
} else if err != nil {
	return err
}
```

Reply:
If you don't care about the wrapping of "failed to create vault" you can also go for

```
vault, err := c.getVaultByName(ctx, c.vaultName)
if errors.Is(err, errVaultDoesNotExist) {
	vault, err = c.vaultClient.Create(ctx, &ngrok.VaultCreate{
		Name:        c.vaultName,
		Description: defaultDescription,
	})
}
if err != nil {
	return err
}
```